src/shogun/modelselection/NewGridSearch.cpp

```diff
@@ -0,0 +1,891 @@
+{
+	m_nodes[name].emplace_back(obj);
+}
+
+ParameterNode::ParameterNode(
+    CSGObject* model, ParameterProperties param_properties)
+{
+	SG_REF(model)
+	m_parent = model;
+
+	for (auto const& param : m_parent->get_params())
+	{
+		// check the param is neither of type CLabels or CFeatures
+		if ((std::type_index(param.second->get_value().type_info()) !=
+		     std::type_index(typeid(CLabels*))) &&
+		    (std::type_index(param.second->get_value().type_info()) !=
+		     std::type_index(typeid(CFeatures*))))
+		{
+			if (auto* value = m_parent->get(param.first, std::nothrow))
+			{
+				auto node = new ParameterNode(value);
+				create_node(param.first, node);
+			}
+		}
+	}
+
+	for (auto const& param : model->get_params(param_properties))
+	{
+		m_param_mapping.insert(
+		    std::make_pair(param.first, param.second->get_value()));
+	}
+}
+
+ParameterNode* ParameterNode::attach(
+    const std::string& param, const std::shared_ptr<ParameterNode>& node)
+{
+	return attach(param, node.get());
+}
+
+ParameterNode*
+ParameterNode::attach(const std::string& param, ParameterNode* node)
+{
+	auto* node_copy = new ParameterNode(*node);
+	if (!set_param_helper(param, node_copy))
+		SG_SERROR("Could not attach %s\n", param.c_str());
+	return this;
+}
+
+ParameterNode*
+ParameterNode::attach(const std::string& param, ParameterNode&& node)
+{
+	if (!set_param_helper(param, &node))
+		SG_SERROR("Could not attach %s\n", param.c_str());
+	return this;
+}
+
+bool ParameterNode::set_param_helper(
+    const std::string& param, ParameterNode* node)
+{
+	auto param_iter = param.find(delimiter);
+
+	if (param_iter != std::string::npos)
+	{
+		return set_param_helper(
+		    param.substr(param_iter + delimiter.size()), node);
+	}
+	else
+	{
+		if (m_parent->has(param))
+		{
+			create_node(param, node);
+			return true;
+		}
+	}
+	return false;
+}
+
+std::string ParameterNode::to_string() const noexcept
+{
+	std::stringstream ss;
+	std::unique_ptr<AnyVisitor> visitor(new ToStringVisitor(&ss));
+	ss << get_name() << "(" << get_parent_name() << "(";
+
+	for (auto param_pair = m_param_mapping.begin();
+	     param_pair != m_param_mapping.end(); ++param_pair)
+	{
+		ss << param_pair->first.c_str() << "=";
+		param_pair->second.visit(visitor.get());
+		if (std::next(param_pair) != m_param_mapping.end())
+			ss << ", ";
+	}
+
+	if (!m_nodes.empty())
+	{
+		if (!m_param_mapping.empty())
+			ss << ", ";
+
+		for (auto node = m_nodes.begin(); node != m_nodes.end(); ++node)
+		{
+			ss << (*node).first << "=";
+			if (node->second.size() > 1)
+			{
+				ss << "[";
+			}
+			for (auto inner_node = node->second.begin();
+			     inner_node != node->second.end(); ++inner_node)
+			{
+				(*inner_node)->to_string_helper(ss, visitor);
+				if (std::next(inner_node) != node->second.end())
+					ss << ", ";
+			}
+			if (node->second.size() > 1)
+			{
+				ss << "]";
+			}
+			if (std::next(node) != m_nodes.end())
+				ss << ", ";
+		}
+	}
+	ss << "))";
```

Comment:
just curious, could you paste an output of this?

Reply:
A node could look like this: 
```
ParameterNode(LibSVM(kernel=[GaussianKernel(log_width=10), PolyKernel(degree=2)]))
```
A GridSearch could look like this:
```
GridSearch(LibSVM(C1=[0.1, 1, 10], C2=[0.1, 1, 10], kernel=[GaussianKernel(log_width=[0.1, 1, 2]), PolyKernel(degree=[1, 2, 3])]))
```

Reply:
sweet! really nice